clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java

```diff
@@ -443,13 +444,13 @@ public boolean isConnectionSetupTimeout(String id, long now) {
     }
 
     /**
-     * Return the Set of nodes whose connection setup has timed out.
+     * Return the List of nodes whose connection setup has timed out.
      * @param now the current time in ms
      */
-    public Set<String> nodesWithConnectionSetupTimeout(long now) {
+    public List<String> nodesWithConnectionSetupTimeout(long now) {
         return connectingNodes.stream()
             .filter(id -> isConnectionSetupTimeout(id, now))
-            .collect(Collectors.toSet());
+            .collect(Collectors.toCollection(LinkedList::new));
```

Comment:
@dajac Won't this create a list regardless of whether there are timed out entries? Let's see whether this addresses @ijuma 's concern on the other PR.

Reply:
@rajinisivaram Yes, it does as an empty list is returned if there are no timed out entries. This was true with the Set as well btw.

Reply:
Why are we using `LinkedList`? It's very rare where it should be used.

Reply:
We only add elements to the List and then iterate over it. A LinkedList seems slightly better for this case.

Reply:
Not at all. This is a perfect case for ArrayList. LinkedList iteration is very slow due to pointer chasing (in comparison).

Reply:
FYI https://twitter.com/joshbloch/status/583813919019573248 :)

Reply:
That’s good to know. I always thought that the iteration was more or less equivalent. I have learnt something today. Let me update that.

Reply:
Done.